### PR TITLE
Make tickets renewable

### LINKED ISSUE #11: Kerberos - renew_lifetime setting make ranger service unable to start
Hello all !

We encountered an issue with the `renew_lifetime = 7d`  setting in `krb5.conf` (_see conf. files below_).
Indeed, when we add first of all this setting in `krb5.conf` (_without update `kdc.conf`_), the ranger service doesn't start (_please see `catalina.out` logs below too_). We hoped that adding the setting `default_principal_flags = +renewable` in `kdc.conf` would make the things working well, but we finally observed the same behaviour.

So as a workaround, we comment the `renew_lifetime = 7d` setting in `krb5.conf` and the ranger service started (_after a manual restart ranger service command_) !

Would you know why this behaviour and how to configure a renewable kerberos token please ?

**krb5.conf.j2**
```
includedir /etc/krb5.conf.d/

[logging]
 default = FILE:/var/log/krb5libs.log
 kdc = FILE:/var/log/krb5kdc.log
 admin_server = FILE:/var/log/kadmind.log

[libdefaults]
 dns_lookup_realm = false
 ticket_lifetime = 24h
 forwardable = true
 rdns = false
 pkinit_anchors = FILE:/etc/pki/tls/certs/ca-bundle.crt
 default_realm = {{ realm }}
 default_ccache_name = /tmp/krb5cc_%{uid}
 canonicalize = true
 renew_lifetime = 7d
 kdc_timeout = 5000
 dns_lookup_kdc = false

[realms]
{{ realm }} = {
  kdc = {{ kerberos_server }}
  admin_server = {{ kerberos_server }}
  default_domain = {{ domain }}
}

[domain_realm]
.{{ domain }} = {{ realm }}
{{ domain }} = {{ realm }}
```

**kdc.conf.j2**
```
includedir /etc/krb5.conf.d/

[libdefaults]
 default_realm = {{ realm }}

[logging]
 default = FILE:/var/log/krb5libs.log
 kdc = FILE:/var/log/krb5kdc.log
 admin_server = FILE:/var/log/kadmind.log

[realms]
{{ realm }} = {
  kdc = {{ kerberos_server }}
  admin_server = {{ kerberos_server }}
  default_domain = {{ domain }}
  default_principal_flags = +renewable
  database_module = openldap_ldapconf
}

[dbdefaults]
ldap_kerberos_container_dn = cn=krbContainer,{{ ldap_suffix }}

[dbmodules]
openldap_ldapconf = {
  db_library = kldap

  # if either of these is false, then the ldap_kdc_dn needs to
  # have write access
  disable_last_success = true
  disable_lockout  = true

  # this object needs to have read rights on
  # the realm container, principal container and realm sub-trees
  ldap_kdc_dn = {{ ldap_admin_dn }}

  # this object needs to have read and write rights on
  # the realm container, principal container and realm sub-trees
  ldap_kadmind_dn = {{ ldap_admin_dn }}

  ldap_service_password_file = {{ ldap_password_file }}
  ldap_servers = ldapi:///
  ldap_conns_per_server = 5
}
```

**catalina.out logs**
```
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
juin 03, 2022 4:00:00 PM org.apache.ranger.server.tomcat.EmbeddedServer start
INFOS: Deriving webapp folder from catalina.base property. folder=/opt/tdp/ranger-2.0.1-TDP-0.1.0-SNAPSHOT-admin/ews/webapp
juin 03, 2022 4:00:00 PM org.apache.ranger.server.tomcat.EmbeddedServer start
INFOS: Webapp file =/opt/tdp/ranger-2.0.1-TDP-0.1.0-SNAPSHOT-admin/ews/webapp, webAppName = /
juin 03, 2022 4:00:00 PM org.apache.ranger.server.tomcat.EmbeddedServer start
INFOS: Adding webapp [/] = path [/opt/tdp/ranger-2.0.1-TDP-0.1.0-SNAPSHOT-admin/ews/webapp] .....
juin 03, 2022 4:00:00 PM org.apache.catalina.core.StandardContext setPath
AVERTISSEMENT: A context path must either be an empty string or start with a '/' and do not end with a '/'. The path [/] does not meet these criteria and has been changed to []
juin 03, 2022 4:00:00 PM org.apache.ranger.server.tomcat.EmbeddedServer start
INFOS: Finished init of webapp [/] = path [/opt/tdp/ranger-2.0.1-TDP-0.1.0-SNAPSHOT-admin/ews/webapp].
juin 03, 2022 4:00:00 PM org.apache.ranger.server.tomcat.EmbeddedServer start
INFOS: Provided Kerberos Credential : Principal = rangeradmin/master-03.applispfref.sipfref.local@BDP01.APPLISPFREF.SIPFREF.LOCAL and Keytab = /etc/security/keytabs/rangeradmin.service.keytab
juin 03, 2022 4:00:00 PM org.apache.ranger.server.tomcat.EmbeddedServer start
GRAVE: Tomcat Server failed to start:java.io.IOException: Login failure for rangeradmin/master-03.applispfref.sipfref.local@BDP01.APPLISPFREF.SIPFREF.LOCAL from keytab /etc/security/keytabs/rangeradmin.service.keytab
java.io.IOException: Login failure for rangeradmin/master-03.applispfref.sipfref.local@BDP01.APPLISPFREF.SIPFREF.LOCAL from keytab /etc/security/keytabs/rangeradmin.service.keytab
        at org.apache.hadoop.security.SecureClientLogin.loginUserFromKeytab(SecureClientLogin.java:71)
        at org.apache.ranger.server.tomcat.EmbeddedServer.start(EmbeddedServer.java:249)
        at org.apache.ranger.server.tomcat.EmbeddedServer.main(EmbeddedServer.java:69)
Caused by: javax.security.auth.login.LoginException: Message stream modified (41)
        at com.sun.security.auth.module.Krb5LoginModule.attemptAuthentication(Krb5LoginModule.java:810)
        at com.sun.security.auth.module.Krb5LoginModule.login(Krb5LoginModule.java:618)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at javax.security.auth.login.LoginContext.invoke(LoginContext.java:755)
        at javax.security.auth.login.LoginContext.access$000(LoginContext.java:195)
        at javax.security.auth.login.LoginContext$4.run(LoginContext.java:682)
        at javax.security.auth.login.LoginContext$4.run(LoginContext.java:680)
        at java.security.AccessController.doPrivileged(Native Method)
        at javax.security.auth.login.LoginContext.invokePriv(LoginContext.java:680)
        at javax.security.auth.login.LoginContext.login(LoginContext.java:587)
        at org.apache.hadoop.security.SecureClientLogin.loginUserFromKeytab(SecureClientLogin.java:68)
        ... 2 more
Caused by: KrbException: Message stream modified (41)
        at sun.security.krb5.KrbKdcRep.check(KrbKdcRep.java:103)
        at sun.security.krb5.KrbAsRep.decrypt(KrbAsRep.java:159)
        at sun.security.krb5.KrbAsRep.decryptUsingKeyTab(KrbAsRep.java:121)
        at sun.security.krb5.KrbAsReqBuilder.resolve(KrbAsReqBuilder.java:310)
        at sun.security.krb5.KrbAsReqBuilder.action(KrbAsReqBuilder.java:498)
        at com.sun.security.auth.module.Krb5LoginModule.attemptAuthentication(Krb5LoginModule.java:782)
```

### PR DESCRIPTION
**Pull request by gboutry**
#### Which issue(s) this PR fixes:

Fixes #11 

#### Additional comments:

Added a `max_renewable_life` parameter to realm configuration.

Moved `renew_lifetime` from `kdc.conf.j2` to `krb5.conf.j2`, since it's a client configuration and not server.

Here's a diff on a principal attributes before and after deploying with this configuration:
```diff
 [root@edge-01 ~]# kadmin -p admin/admin -w admin123 -q "getprinc tdp_user"
 Authenticating as principal admin/admin with password.
 Principal: tdp_user@REALM.TDP
 Expiration date: [never]
 Last password change: Wed Aug 24 13:13:01 UTC 2022
 Password expiration date: [never]
 Maximum ticket life: 1 day 00:00:00
-Maximum renewable life: 0 days 00:00:00
+Maximum renewable life: 7 days 00:00:00
 Last modified: Wed Aug 24 13:13:01 UTC 2022 (admin/admin@REALM.TDP)
 Last successful authentication: [never]
 Last failed authentication: [never]
 Failed password attempts: 0
 Number of keys: 2
 Key: vno 1, aes256-cts-hmac-sha1-96
 Key: vno 1, aes128-cts-hmac-sha1-96
 MKey: vno 1
 Attributes:
 Policy: [none]
```

To make clear that you license your contribution under
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0) and that you give permission to [TOSIT](https://www.tosit.io/),
you have to acknowledge this by using the following check-box.

 - [X] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [X] I hereby agree to grant [TOSIT](https://www.tosit.io/) a copyright license to use my contributions.

### CODE REVIEW

**Review by rpignolet — APPROVED** [review]

File: roles/ldap_kerberos/templates/krb5.conf.j2 [hunk 1]
```diff
@@ -8,6 +8,7 @@ includedir /etc/krb5.conf.d/
 [libdefaults]
   dns_lookup_realm = false
   ticket_lifetime = 24h
+  renew_lifetime = 7d
```
Comment (nschung) [inline roles/ldap_kerberos/templates/krb5.conf.j2:11]: The effective ticket renew lifetime is the min of renew_lifetime and max_renewable_life. Should theses 2 variables be synchronized?

Comment (gboutry) [reply roles/ldap_kerberos/templates/krb5.conf.j2:11]: We don't really care here, the most important is that the tickets are renewable. It's not a requirement for them to be synchronized.

Comment (nschung) [reply roles/ldap_kerberos/templates/krb5.conf.j2:11]: As max_renewable_life is variablized, ticket_lifetime and renew_lifetime should be settled in the same way . A comment for the guideline utilization of renew_lifetime and max_renewable_life will help for a  better control of ticket renewal lifetime.

Comment (gboutry) [reply roles/ldap_kerberos/templates/krb5.conf.j2]: Ok, I added templatisation of the `renew_lifetime` parameter, added some information in readme about it.

Note that I did not variabilize `ticket_lifetime` as it is not in the scope of this PR.

Comment (nschung) [reply roles/ldap_kerberos/templates/krb5.conf.j2]: It's good for me.

**Review by nschung — APPROVED** [review]

File: roles/ldap_kerberos/templates/krb5.conf.j2 (hunk 1 above)
Comment (gboutry) [reply roles/ldap_kerberos/templates/krb5.conf.j2]: Thanks, I'm going to merge then

**Pull request merged**